### PR TITLE
Worker client draft

### PR DESCRIPTION
**Pull request by nikolatzotchev**
I just wanted to ask if I'm understanding correctly everything.
The WorkerProxy(not sure if it should be called that way) should be invoked from the ssh.
Worker just starts everything and creates a docker client (that the commands will use in there execute() method).

### CODE REVIEW

File: src/ja/common/response_message.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+"""
+This module provides the interface which all respond implement.
```
Comment (ammen99) [inline src/ja/common/response_message.py]: all respond? Maybe you meants responses?

File: src/ja/common/response_message.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+"""
+from ja.common.message import Message
+
+
+class RespondMessage(Message):
+    """
+    A base class for all response commands
+    A respond message is a message
```
Comment (ammen99) [inline src/ja/common/response_message.py]: "A message is a message" anyway:)

File: src/ja/worker/worker_client.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+class Worker:
+    """
+    The main class on the Worker Client
+    """
+
+    def __init__(self) -> None:
+        """!
+        Initializes the Worker, here the SSHConnection with the Server should
+        take place, initializing the DockerClient as well as starting the event loop in DockerEvents
+         that is going to be notifying the Server when necessary.
```
Comment (ammen99) [inline src/ja/worker/worker_client.py]: Why the space in the beginning of the line?

File: src/ja/worker/worker_client.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        """
+
+    def send_respond_message(self, msg: RespondMessage) -> None:
+        """!
+        This method will take a ResponseMessage and will send it to the Central Server.
+        @param msg The @RespondMessage
+        """
+
+
+class DockerEvents:
```
Comment (ammen99) [inline src/ja/worker/worker_client.py]: Actually I am not sure how to implement this class. I imagined we'd have some event loop and poll an fd or something?

Comment (nikolatzotchev) [reply src/ja/worker/worker_client.py]: I tough we can use https://docker-py.readthedocs.io/en/stable/client.html#docker.client.DockerClient.events
and just filter for example when e container finishes.

Comment (ammen99) [reply src/ja/worker/worker_client.py]: Didn't know the API looks like that.
Anyway I realized the model depends on what we decide (server asks for updates or wc notifies server)

File: src/ja/common/message/worker.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    """
+    A base class for all worker commands. A worker command is an action which
+    is sent by the central server and performed on the worker client. WorkerCommands
+    typically execute some kind of docker commands on the worker client.
+    """
+
+
+class WorkerResponse(Response):
+    """
+    A base class for all worker client responses. A server response is a Response
```
Comment (JohannesGaessler) [inline src/ja/common/message/worker.py]: A server response -> A worker response

File: src/ja/worker/docker_events.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+"""
+This module is responsible for the monitoring of the docker containers running on the worker machine
+"""
+
+
+class DockerEvents:
+    """
+    Monitors the containers no the current machine
```
Comment (JohannesGaessler) [inline src/ja/worker/docker_events.py]: no -> on

File: src/ja/worker/proxy/remote.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+LocalCentralServer is a single-class command line program.
+
+It queues the command on the worker socket.
+
+The commands are received on the standard input, and the worker response
+will be printed on the standard output.
+"""
+
+
+class LocalCentralServer:
```
Comment (JohannesGaessler) [inline src/ja/worker/proxy/remote.py]: This should be called WorkerRemote and be a sub-class of the abstract Remote class in ja.common.proxy.remote

**Comment by nikolatzotchev** [comment]
What do you mean by 

> You still need to specify the interfaces of these classes.

Do mean it like adding interfaces of the classes, or like adding more methods to the classes.

**Comment by JohannesGaessler** [comment]
@nikolatzotchev I mean that you need to add method stubs.

**Comment by nikolatzotchev** [comment]
But only public methods right, private methods will be made lather right?

**Comment by JohannesGaessler** [comment]
Only public methods.

**Comment by nikolatzotchev** [comment]
I think it will be best to give the execute command of the Worker commands the Worker instance as argument, that way the server can be able to control basically everything on the worker.
The DockerEvents class will use the [docker event](https://docs.docker.com/engine/reference/commandline/events/) command with a filter to get events only on specific criteria.
For python I think it will be best to [Docker SDK for Python ](https://docker-py.readthedocs.io/en/stable/index.html). I'm not so well informed about UNIX sockets (I'm reading some stuff right now) and I don't know if we will need a special class for them.

**Comment by ammen99** [comment]
Is this a draft of is this ready for review&merge? (I see @JohannesGaessler already approved)

**Comment by JohannesGaessler** [comment]
It seems that was a mistake on my part.
I was just going through the changes to see if they seem reasonable and I did not consider that there might still be things missing.

**Comment by nikolatzotchev** [comment]
I totally forgot to remove the draft thing sorry.

File: src/ja/worker/docker_events.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    def __init__(self) -> None:
+        """!
+        Starts the event loop
+        that will be monitoring the state of the containers.
+        """
+
+    def change_filters(self) -> None:
+        """!
+        Change on witch events the server is going to be notified
```
Comment (ammen99) [inline src/ja/worker/docker_events.py]: witch -> which

Comment (ammen99) [reply src/ja/worker/docker_events.py]: Also, how is the change done? Should there be a parameter describing what events to listen for?

File: src/ja/worker/docker_events.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    Monitors the containers on the current machine
+    """
+
+    def __init__(self) -> None:
+        """!
+        Starts the event loop
+        that will be monitoring the state of the containers.
+        """
+
+    def change_filters(self, filters: dict) -> None:
```
Comment (ammen99) [inline src/ja/worker/docker_events.py]: There is no description of the filters parameter, I can't see how to use it.

File: src/ja/worker/docker_events.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        """!
+        Change on which events the server is going to be notified
+        :param filters: filter the events by event time, container or image
```
Comment (ammen99) [inline src/ja/worker/docker_events.py]: we use `@` in front in the other places (see my and @JohannesGaessler 's pull requests)

**Pull request merged**